Content.Server/Psionics/PsionicsSystem.cs

```diff
@@ -105,40 +104,56 @@ private void OnStamHit(EntityUid uid, AntiPsionicWeaponComponent component, Take
                 args.FlatModifier += component.PsychicStaminaDamage;
         }
 
-        public void RollPsionics(EntityUid uid, PotentialPsionicComponent component, bool applyGlimmer = true, float multiplier = 1f)
+        public void RollPsionics(EntityUid uid, PsionicComponent component, bool applyGlimmer = true, float rollEventMultiplier = 1f)
         {
-            if (HasComp<PsionicComponent>(uid)
-                || !_cfg.GetCVar(CCVars.PsionicRollsEnabled))
+            if (!_cfg.GetCVar(CCVars.PsionicRollsEnabled))
                 return;
 
-            var chance = component.Chance;
-            var warn = true;
-            if (TryComp<PsionicBonusChanceComponent>(uid, out var bonus))
-            {
-                chance *= bonus.Multiplier;
-                chance += bonus.FlatBonus;
-                warn = bonus.Warn;
-            }
+            // Calculate the initial odds based on the innate potential
+            var baselineChance = component.Chance
+                * component.PowerRollMultiplier
+                + component.PowerRollFlatBonus;
 
-            if (applyGlimmer)
-                chance += ((float) _glimmerSystem.Glimmer / 1000);
+            // Increase the initial odds based on Glimmer.
+            // TODO: Change this equation when I do my Glimmer Refactor
+            baselineChance += applyGlimmer
+                ? (float) _glimmerSystem.Glimmer / 1000 //Convert from Glimmer to %chance
+                : 0;
 
-            chance *= multiplier;
+            // Certain sources of power rolls provide their own multiplier.
+            baselineChance *= rollEventMultiplier;
 
-            chance = Math.Clamp(chance, 0, 1);
+            // Ask if the Roller has any other effects to contribute, such as Traits.
+            var ev = new OnRollPsionicsEvent(uid, baselineChance);
+            RaiseLocalEvent(uid, ref ev);
 
-            if (_random.Prob(chance))
-                _psionicAbilitiesSystem.AddPsionics(uid, warn);
+            if (_random.Prob(ev.BaselineChance))
+                _psionicAbilitiesSystem.AddPsionics(uid);
         }
 
-        public void RerollPsionics(EntityUid uid, PotentialPsionicComponent? psionic = null, float bonusMuliplier = 1f)
+        public void RerollPsionics(EntityUid uid, PsionicComponent? psionic = null, float bonusMuliplier = 1f)
         {
             if (!Resolve(uid, ref psionic, false)
                 || psionic.Rerolled)
                 return;
 
-            RollPsionics(uid, psionic, multiplier: bonusMuliplier);
+            RollPsionics(uid, psionic, true, bonusMuliplier);
             psionic.Rerolled = true;
         }
     }
 }
+
+/// <summary>
```

Comment:
are events not usually in their own files?

Reply:
Done. There is now a PsionicsSystem.Events.cs file.